core/client/hammer/hammer.go

```diff
@@ -0,0 +1,196 @@
+)
+
+// CustomDial returns a grpc connection.
+var CustomDial func(ctx context.Context, addr string, opts ...grpc.DialOption) (pb.KeyTransparencyClient, error)
+
+// Hammer represents a single run of the hammer.
+type Hammer struct {
+	workers int
+
+	timeout time.Duration
+	ktAddr  string
+	appID   string
+	config  *pb.Domain
+
+	signers        []*tink.KeysetHandle
+	authorizedKeys *tinkpb.Keyset
+}
+
+// New returns a new hammer job
+func New(ctx context.Context, ktAddr, domainID string, timeout time.Duration, keyset *tink.KeysetHandle) (*Hammer, error) {
+	client, err := CustomDial(ctx, ktAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	config, err := client.GetDomain(ctx, &pb.GetDomainRequest{DomainId: domainID})
+	if err != nil {
+		return nil, err
+	}
+
+	authorizedKeys, err := keyset.GetPublicKeysetHandle()
+	if err != nil {
+		return nil, fmt.Errorf("keyset.GetPublicKeysetHandle() failed: %v", err)
+	}
+
+	return &Hammer{
+		timeout: timeout,
+		ktAddr:  ktAddr,
+		appID:   fmt.Sprintf("hammer_%v", time.Now().Format("2006-01-02/15:04:05")),
+		config:  config,
+
+		signers:        []*tink.KeysetHandle{keyset},
+		authorizedKeys: authorizedKeys.Keyset(),
+	}, nil
+}
+
+// Run runs operationCount with up to maxWorkers over ramp duration.
+func (h *Hammer) Run(ctx context.Context, operationCount, maxWorkers int, ramp time.Duration) {
+	jobs := gen(ctx, operationCount)
+	outputs := h.launchWorkers(ctx, jobs, maxWorkers, ramp)
+	h.printStats(1*time.Second, outputs)
+}
+
+// launchWorkers gradually adds workers to the worker pool over ramp up to workers.
+func (h *Hammer) launchWorkers(ctx context.Context, jobs <-chan int, workers int, ramp time.Duration) <-chan error {
+	outputs := make(chan error)
+
+	go func() {
+		rampTicker := time.NewTicker(ramp / time.Duration(workers))
+		var wg sync.WaitGroup
+		for ; h.workers < workers; h.workers++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				h.worker(ctx, jobs, outputs)
+			}()
+			<-rampTicker.C
+		}
+		rampTicker.Stop()
+
+		wg.Wait()
+		close(outputs)
+	}()
+
+	return outputs
+}
+
+// gen produces a channel of integers from 0 up to count-1 and then closes.
+func gen(ctx context.Context, count int) <-chan int {
+	jobs := make(chan int)
+	go func() {
+		defer close(jobs)
+		for i := 0; i < count; i++ {
+			select {
+			case <-ctx.Done():
+				return
+			case jobs <- i:
+			}
+		}
+	}()
+	return jobs
+}
+
+// printStats prints stats repeatedly every period until outputs closes.
+func (h *Hammer) printStats(period time.Duration, outputs <-chan error) {
+	startTime := time.Now()
+	var totalRequests float64
+
+	// Print stats asyncronously until refresh closes.
+	refresh := time.NewTicker(period)
+	stop := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-refresh.C:
+				rate := totalRequests / time.Since(startTime).Seconds()
+				glog.Infof("QPS: %v, \tRequests: %v, \tWorkers: %v", rate, totalRequests, h.workers)
+			case <-stop:
```

Comment:
Shouldn't you write to `stop` channel to make this `case` happen?

Reply:
`close(stop)` causes the channel return immediately